katdal/applycal.py

```diff
@@ -92,42 +86,50 @@ def complex_interp(x, xi, yi, left=None, right=None):
     return y.astype(yi.dtype)
 
 
-def has_cal_product(cache, attrs, product):
-    """Check if calibration solution `product` is available in `cache`."""
-    key = 'cal_product_' + product
-    try:
-        parts = int(attrs[key + '_parts'])
-    except KeyError:
-        return key in cache
-    else:
-        return any(key + str(part) in cache for part in range(parts))
+def _parse_cal_product(cal_product):
+    """Split `cal_product` into `cal_stream` and `product_type` parts."""
+    fields = cal_product.rsplit('.', 1)
+    product_type = fields.pop()
+    cal_stream = fields.pop() if fields else 'cal'
+    return cal_stream, product_type
 
 
-def get_cal_product(cache, attrs, product):
-    """Extract calibration solution `product` from `cache` as a sensor.
+def get_cal_product(cache, attrs, product_type, cal_stream='cal'):
```

Comment:
Perhaps `cal_stream` shouldn't have a default if the default is wrong for every case except unit tests.

Reply:
Agreed. The previous comment is also True about `_parse_cal_product`. I'll rework the defaults.